tests/unit/test_multicast_listeners_table.cpp

```diff
@@ -0,0 +1,211 @@
+                 "Add should fail");
+    VerifyOrQuit(table.Add(static_cast<const Ip6::Address &>(MA301), TimerMilli::GetNow()) == OT_ERROR_INVALID_ARGS,
+                 "Add should fail");
+
+    // Expire should expire outdated Listeners
+    sNow = 2;
+    table.Expire();
+    VerifyOrQuit(table.Count() == 0, "Table count is wrong");
+
+    // Add different addresses until the table is full
+    sNow = 10;
+    for (uint16_t i = 0; i < OPENTHREAD_CONFIG_MAX_MULTICAST_LISTENERS; i++)
+    {
+        Ip6::Address address;
+
+        address                = static_cast<const Ip6::Address &>(MA401);
+        address.mFields.m16[7] = HostSwap16(i);
+
+        SuccessOrQuit(table.Add(address, TimerMilli::GetNow() + i), "Add failed");
+        VerifyOrQuit(table.Count() == i + 1, "Table count is wrong");
+    }
+
+    // Now the table is full, we can't add more addresses
+    VerifyOrQuit(table.Add(static_cast<const Ip6::Address &>(MA501), TimerMilli::GetNow()) == OT_ERROR_NO_BUFS,
+                 "Add should fail");
+
+    // Expire one Listener at a time
+    for (uint16_t i = 0; i < OPENTHREAD_CONFIG_MAX_MULTICAST_LISTENERS; i++)
+    {
+        table.Expire();
+        VerifyOrQuit(table.Count() == OPENTHREAD_CONFIG_MAX_MULTICAST_LISTENERS - i - 1, "Table count is wrong");
+
+        sNow += 1;
+    }
+
+    // Now the table should be empty
+    VerifyOrQuit(table.Count() == 0, "Table count is wrong");
+
+    // Now test the APIs
+    testMulticastListenersTableAPIs(sInstance);
+
+    // Do some fuzzy test
+    for (uint16_t i = 0; i < 10000; i++)
+    {
+        Ip6::Address address;
+        sNow += 10;
+
+        table.Expire();
+
+        for (MulticastListenersTable::Listener &listener : table.Iterate())
+        {
+            OT_ASSERT(listener.GetAddress().IsMulticastLargerThanRealmLocal());
+            OT_ASSERT(listener.GetExpireTime() > TimerMilli::GetNow());
+        }
+
+        address = static_cast<const Ip6::Address &>(MA401);
+
+        address.mFields.m16[7] = Random::NonCrypto::GetUint16InRange(1, 1000);
+        IgnoreError(table.Add(address, TimerMilli::GetNow() + Random::NonCrypto::GetUint32InRange(1, 100)));
+
+        address.mFields.m16[7] = Random::NonCrypto::GetUint16InRange(1, 1000);
+        if (Random::NonCrypto::GetUint16InRange(0, 2) == 0)
+        {
+            table.Remove(address);
+        }
+    }
+}
+
+void testMulticastListenersTableAPIs(Instance *aInstance)
+{
+    otBackboneRouterMulticastListenerIterator iter = OT_BACKBONE_ROUTER_MULTICAST_LISTENER_ITERATOR_INIT;
+    otBackboneRouterMulticastListenerInfo     info;
+    size_t                                    table_size = 0;
+
+    while (otBackboneRouterMulticastListenerGetNext(aInstance, &iter, &info) == OT_ERROR_NONE)
+    {
+        VerifyOrQuit(false, "Table should be empty");
+    }
+
+    SuccessOrQuit(otBackboneRouterMulticastListenerAdd(aInstance, &MA401, 30000), "Add failed");
```

Comment:
These APIs are only defined under OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE.

I think should add a `#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE` around this block of code (to avoid linker error).

Reply:
added `#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE`